### PR TITLE
Julia Client & C API Extensions

### LINKED ISSUE #3055: When non utf8 data is passed to `duckdb_bind_varchar_length` it will crash the process instead of returning `DuckDBError`
#### What happens?

When you use `duckdb_bind_varchar_length`, and pass non utf8 data (by accident), it crashes the process, instead of returning `DuckDbError`.

I know you should use blobs for arbitrary binary data, but you don't always have control over input in strings. Would it be possible to catch the exception below and return an error?

Output: 
```
libc++abi: terminating with uncaught exception of type duckdb::Exception: String value is not valid UTF8
make: *** [test] Abort trap: 6
```

#### To Reproduce

Call `duckdb_varchar_length` with as input `[0, 40, 41]`, and length `3`.

#### Environment (please complete the following information):
 - OS: macOS
 - DuckDB Version: 0.3.2
 - DuckDB Client: WIP erlang interface.

#### Before Submitting

Tried the latest libduckdb.zip from GH

### FYI

Testcase looks like this:

```erlang
bind_varchar_test() ->
    {ok, Db} = educkdb:open(":memory:"),
    {ok, Conn} = educkdb:connect(Db),

    {ok, [], []} = q(Conn, "create table test(a varchar(10), b varchar(200));"),
    {ok, Insert} = educkdb:prepare(Conn, "insert into test values($1, $2);"),

    ok = educkdb:bind_varchar(Insert, 1, "hello"),
    ok = educkdb:bind_varchar(Insert, 2, "world"),

    {ok, _, [[1]]} = x(Insert),

    ok = educkdb:bind_varchar(Insert, 1, <<"😀"/utf8>>),
    ok = educkdb:bind_varchar(Insert, 2, <<0, "()">>),   % <-- crash here

    {ok, _, [[1]]} = x(Insert),

    {ok, _, [
             [<<"hello">>, <<"world">>],
             [<<"😀"/utf8>>, <<"1234567890">>]
            ]} = q(Conn, "select * from test order by a"),
    ok.
```

### PR DESCRIPTION
**Pull request by Mytherin**
This PR adds a Julia client to DuckDB. The Julia client is  built on top of [the original Julia client](https://github.com/kimmolinna/DuckDB.jl) by @kimmolinna, but extended in various ways:

* DBInterface.jl is now fully supported (see [test_basic_queries.jl](https://github.com/Mytherin/duckdb/blob/julia/tools/juliapkg/test/test_basic_queries.jl) and [test_sqlite.jl](https://github.com/Mytherin/duckdb/blob/julia/tools/juliapkg/test/test_sqlite.jl))
* Support for [prepared statements](https://github.com/Mytherin/duckdb/blob/julia/tools/juliapkg/test/test_prepare.jl)
* [All data types](https://github.com/Mytherin/duckdb/blob/julia/tools/juliapkg/test/test_all_types.jl) are supported in the result fetch, including composite and complex data types (lists, structs, etc)
* [Scanning data frames](https://github.com/Mytherin/duckdb/blob/julia/tools/juliapkg/test/test_df_scan.jl) is supported
* Support for creating [table functions](https://github.com/Mytherin/duckdb/blob/julia/tools/juliapkg/test/test_table_function.jl) and [replacement scans](https://github.com/Mytherin/duckdb/blob/julia/tools/juliapkg/test/test_replacement_scan.jl) in pure Julia
* Support for [configuration parameters](https://github.com/Mytherin/duckdb/blob/julia/tools/juliapkg/test/test_config.jl)

### C API Extensions

The Julia client is built on top of the C interface using Julia's (excellent) C FFI support, see the [API bindings](https://github.com/Mytherin/duckdb/blob/julia/tools/juliapkg/src/api.jl).

As part of making an efficient integration, the C API has been extended in various ways. In particular, the C interface has been extended to reflect the internal C++ structures more closely. The classes `duckdb_logical_type`, `duckdb_data_chunk`, `duckdb_vector`, `duckdb_value` have been added, which are direct wrappers for their C++ counterparts. Functions have been added to operate on these. See [the documentation](https://duckdb.org/docs/api/c/data_chunk).

Using these structures is the fastest (and therefore recommended) way of using various parts of the C API, such as reading results using `duckdb_result_get_chunk` or appending data to an appender using `duckdb_append_data_chunk`.

As part of this PR we add support for structurally reading all the DuckDB types to the C API as well, including composite types such as lists, structs and maps, and complex types such as enums and decimals. This support is only added to the data chunk API, however. While the old result-set API still exists and works, it has not been extended to support types it did not previously. 

In addition, this PR also adds support for defining table functions and replacement scans to the C API. Table functions internally require the use of the data chunk API as well.

Also fixes #3055 and #2733

### CODE REVIEW

**Comment by kimmolinna** [comment]
This is really great. It was my pleasure to be a kickstarter for this. Julia and DuckDB are a really powerful combo. The Academic world produce performant tools too.

**Review by samansmink — COMMENTED** [review]
Only looked at the CI, added 2 comments

File: .github/workflows/Julia.yml [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+       - 'data/**'
+       - 'examples/**'
+       - 'test/**'
+       - 'tools/nodejs/**'
+       - 'tools/pythonpkg/**'
+       - 'tools/rpkg/**'
+       - '.github/workflows/**'
+       - '!.github/workflows/Julia.yml'
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}-${{ github.head_ref || '' }}-${{ github.base_ref || '' }}
```
[the remark this answers is not in the document]
Comment (Mytherin) [reply .github/workflows/Julia.yml:18]: That's not on purpose, thanks!

File: .github/workflows/Julia.yml [hunk 2]
```diff
@@ -0,0 +1,78 @@
+name: Julia
+on:
+   push:
+     paths-ignore:
+       - '**.md'
+   pull_request:
+     paths-ignore:
+       - '**.md'
+       - 'data/**'
```
[the remark this answers is not in the document]
Comment (Mytherin) [reply .github/workflows/Julia.yml:9]: Yes, good point. Have removed it again.

**Pull request merged**